middleware/kubernetes/autopath.go

```diff
@@ -0,0 +1,82 @@
+package kubernetes
+
+import "github.com/miekg/dns"
+
+// AutoPathWriter implements a ResponseWriter that also does the following:
+// * reverts question section of a packet to its original state.
+//   This is done to avoid the 'Question section mismatch:' error in client.
+// * Defers write to the client if the response code is NXDOMAIN.  This is needed
+//   to enable further search path probing if a search was not sucessful.
+// * Allow forced write to client regardless of response code.  This is needed to
+//   write the packet to the client if the final search in the path fails to
+//   produce results.
+// * Overwrites response code with AutoPathWriter.Rcode if the response code
+//   is NXDOMAIN (NameError).  This is needed to support the AutoPath.OnNXDOMAIN
+//   function, which returns a NOERROR to client instead of NXDOMAIN if the final
+//   search in the path fails to produce results.
+
+type AutoPathWriter struct {
+	dns.ResponseWriter
+	original dns.Question
+	Rcode    int
+	Sent     bool
+}
+
+// NewAutoPathWriter returns a pointer to a new AutoPathWriter
+func NewAutoPathWriter(w dns.ResponseWriter, r *dns.Msg) *AutoPathWriter {
+	return &AutoPathWriter{
+		ResponseWriter: w,
+		original:       r.Question[0],
+		Rcode:          dns.RcodeSuccess,
+	}
+}
+
+func (apw *AutoPathWriter) WriteMsg(res *dns.Msg) error {
+	return apw.OverrideMsg(res, false)
+}
+
+// Force the write to client regardless of response code
+func (apw *AutoPathWriter) ForceWriteMsg(res *dns.Msg) error {
+	return apw.OverrideMsg(res, true)
+}
+
+// OverrideMsg records the status code and calls the
+// underlying ResponseWriter's WriteMsg method unless
+// the write is deferred, or force = true.
+func (apw *AutoPathWriter) OverrideMsg(res *dns.Msg, force bool) error {
```

Comment:
There is WriteMsg, and ForceWriteMsg.  The first calls OverrideMsg with force=false. the second calls OverrideMsg with force=true.

I should unexpose OverrideMsg, might make is clearer.